README.md
Syro
====

Simple router for web applications.

Community
---------

Meet us on IRC: [#syro](irc://chat.freenode.net/#syro) on
[freenode.net](http://freenode.net/).

Description
-----------

Syro is a very simple router for web applications. It was created
in the tradition of libraries like [Rum][rum] and [Cuba][cuba], but
it promotes a less flexible usage pattern. The design is inspired
by the way some Cuba applications are architected: modularity is
encouraged and sub-applications can be dispatched without any
significant performance overhead.

Check the [website][syro] for more information, and follow the
[tutorial][tutorial] for a step by step introduction.

[rum]: http://github.com/chneukirchen/rum
[cuba]: http://cuba.is
[syro]: http://soveran.github.io/syro/
[tutorial]: http://files.soveran.com/syro/

Usage
-----

An example of a modular application would look like this:

```ruby
Admin = Syro.new do
  get do
    res.write "Hello from admin!"
  end
end

App = Syro.new do
  on "admin" do
    run(Admin)
  end
end
```

The block is evaluated in a sandbox where the following methods are
available: `env`, `req`, `res`, `path`, `inbox`, `call`, `run`,
`halt`, `consume`, `capture`, `root?` `match`, `default`, `on`,
`root`,`get`, `put`, `head`, `post`, `patch`, `delete` and `options`.
Three other methods are available for customizations: `default_headers`,
`request_class` and `response_class`.

As a recommendation, user created variables should be instance
variables. That way they won't mix with the API methods defined in
the sandbox. All the internal instance variables defined by Syro
are prefixed by `syro_`, like in `@syro_inbox`.

API
---

`env`: Environment variables for the request.

`req`: Helper object for accessing the request variables. It's an
instance of `Rack::Request`.

`res`: Helper object for creating the response. It's an instance
of `Syro::Response`.

`path`: Helper object that tracks the previous and current path.

`inbox`: Hash with captures and potentially other variables local
to the request.

`call`: Entry point for the application. It receives the environment
and optionally an inbox.

`run`: Runs a sub app, and accepts an inbox as an optional second
argument.

`halt`: Terminates the request. It receives an array with the
response as per Rack's specification.

`consume`: Match and consume a path segment.

`capture`: Match and capture a path segment. The value is stored in
the inbox.

`root?`: Returns true if the path yet to be consumed is empty.

`match`: Receives a String, a Symbol or a boolean, and returns true
if it matches the request.

`default`: Receives a block that will be executed inconditionally.

`on`: Receives a value to be matched, and a block that will be
executed only if the request is matched.

`root`: Receives a block and calls it only if `root?` is true.

`get`: Receives a block and calls it only if `root?` and `req.get?` are
true.

`put`: Receives a block and calls it only if `root?` and `req.put?` are
true.

`head`: Receives a block and calls it only if `root?` and `req.head?`
are true.

`post`: Receives a block and calls it only if `root?` and `req.post?`
are true.

`patch`: Receives a block and calls it only if `root?` and `req.patch?`
are true.

`delete`: Receives a block and calls it only if `root?` and `req.delete?`
are true.

`options`: Receives a block and calls it only if `root?` and
`req.options?` are true.

Decks
-----

The sandbox where the application is evaluated is an instance of
`Syro::Deck`, and it provides the API described earlier. You can
define your own `Deck` and pass it to the `Syro` constructor. All
the methods defined in there will be accessible from your routes.
Here's an example:

```ruby
class TextualDeck < Syro::Deck
  def text(str)
    res[Rack::CONTENT_TYPE] = "text/plain"
    res.write(str)
  end
end

App = Syro.new(TextualDeck) do
  get do
    text("hello world")
  end
end
```

The example is simple enough to showcase the concept, but maybe too
simple to be meaningful. The idea is that you can create your own
specialized decks and reuse them in different applications. You can
also define modules and later include them in your decks: for
example, you can write modules for rendering or serializing data,
and then you can combine those modules in your custom decks.

Examples
--------

In the following examples, the response string represents
the request path that was sent.

```ruby
App = Syro.new do
  get do
    res.write "GET /"
  end

  post do
    res.write "POST /"
  end

  on "users" do
    on :id do

      # Captured values go to the inbox
      @user = User[inbox[:id]]

      get do
        res.write "GET /users/42"
      end

      put do
        res.write "PUT /users/42"
      end

      patch do
        res.write "PATCH /users/42"
      end

      delete do
        res.write "DELETE /users/42"
      end
    end

    get do
      res.write "GET /users"
    end

    post do
      res.write "POST /users"
    end
  end
end
```

Matches
-------

The `on` method can receive a `String` to perform path matches; a
`Symbol` to perform path captures; and a boolean to match any true
values.

Each time `on` matches or captures a segment of the PATH, that part
of the path is consumed. The current and previous paths can be
queried by calling `prev` and `curr` on the `path` object: `path.prev`
returns the part of the path already consumed, and `path.curr`
provides the current version of the path.

Any expression that evaluates to a boolean can also be used as a
matcher.  For example, a common pattern is to follow some route
only if a user is authenticated. That can be accomplished with
`on(authenticated(User))`. That example assumes there's a method
called `authenticated` that returns true or false depending on
whether or not an instance of `User` is authenticated. As a side
note, [Shield][shield] is a library that provides just that.

[shield]: https://github.com/cyx/shield

Captures
--------

When a symbol is provided, `on` will try to consume a segment of
the path. A segment is defined as any sequence of characters after
a slash and until either another slash or the end of the string.
The captured value is stored in the `inbox` hash under the key that
was provided as the argument to `on`. For example, after a call to
`on(:user_id)`, the value for the segment will be stored at
`inbox[:user_id]`. When mounting an application called `Users` with
the command `run(Users)`, an inbox can be provided as the second
argument: `run(Users, inbox)`. That allows apps to share previous
captures.

Status code
-----------

By default the status code is set to `404`. If both path and request
method are matched, the status is automatically changed to `200`.
You can change the status code by assigning a number to `res.status`,
for example:

```ruby
post do
  ...
  res.status = 201
end
```

Content type
------------

There's no default value for the content type header, but there's
a handy way of setting the desired value.

In order to write the body of the response, the `res.write` method
is used:

```ruby
res.write "hello world"
```

It has the drawback of leaving the `Content-Type` header empty.
Three alternative methods are provided, and more can be added by
using custom Decks.

Setting the Content-Type as `"text/plain"`:

```ruby
res.text "hello world"
```

Setting the Content-Type as `"text/html"`:

```ruby
res.html "hello world"
```

Setting the Content-Type as `"application/json"`:

```ruby
res.json "hello world"
```

Note that aside from writing the response body and setting the value
for the Content-Type header, no encoding or serialization takes
place. If you want to return a JSON encoded response, make sure to
encode the objects yourself (i.e., `res.json JSON.dump(...)`).

Security
--------

There are no security features built into this routing library. A
framework using this library should implement the security layer.

Rendering
---------

There are no rendering features built into this routing library. A
framework that uses this routing library can easily implement helpers
for rendering.

Middleware
----------

Syro doesn't support Rack middleware out of the box. If you need them,
just use `Rack::Builder`:

```ruby
App = Rack::Builder.new do
  use Rack::Session::Cookie, secret: "..."

  run Syro.new do
    get do
      res.write("Hello, world")
    end
  end
end
```

Trivia
------

An initial idea was to release a new version of [Cuba](http://cuba.is)
that broke backward compatibility, but in the end my friends suggested
to release this as a separate library. In the future, some ideas
of this library could be included in Cuba as well.

Installation
------------

```
$ gem install syro
```


CHANGELOG
3.0.0

* Change how status codes are set.
* Change how the Content-Type is set.
* Add Content-Type helpers to Syro::Response.

2.2.0

* Change gemspec to allow Rack 2.

2.1.2

* Always set default content type if body is set.

2.1.1

* Change gemspec to accept pre-release versions of Rack 2

2.1.0

* Add matchers for head and options

2.0.0

* No changes since 2.0.0.rc1

2.0.0.rc1

* Add consume and capture primitives

* Add default matcher

* Revert yield of captured value (was added in 1.0.0)

1.1.1

* Small internal refactoring

1.1.0

* Move Deck methods to the Deck::API module

* Add support for custom response and request classes

1.0.0

* Yield captures as an alternative API

* Add support for default headers in response

0.0.8

* Reset PATH_INFO and SCRIPT_NAME after running a subapp

0.0.7

* Change run so that it accepts Rack applications

0.0.6

* Add put request method

0.0.5

* Add explicit Rack version dependency (@luislavena)

0.0.4

* Add concept of Decks

0.0.3

* Remove request method overriding (@larrylv)

0.0.2

* Defer method comparisons to Rack::Request (@pote)


test/all.rb
require "json"

class RackApp
  def call(env)
    [200, {"Content-Type" => "text/html"}, ["GET /rack"]]
  end
end

class MarkdownDeck < Syro::Deck
  def markdown(str)
    res[Rack::CONTENT_TYPE] = "text/markdown"
    res.write(str)
  end
end

class DefaultHeaders < Syro::Deck
  def default_headers
    { Rack::CONTENT_TYPE => "text/html" }
  end
end

class CustomRequestAndResponse < Syro::Deck
  class JSONRequest < Rack::Request
    def params
      JSON.parse(body.read)
    end
  end

  class JSONResponse < Syro::Response
    def write(s)
      super(JSON.generate(s))
    end
  end

  def request_class
    JSONRequest
  end

  def response_class
    JSONResponse
  end
end

markdown = Syro.new(MarkdownDeck) do
  get do
    markdown("GET /markdown")
  end
end

default_headers = Syro.new(DefaultHeaders) do end

json = Syro.new(CustomRequestAndResponse) do
  root do
    params = req.params

    res.write(params)
  end
end

admin = Syro.new do
  get do
    res.write("GET /admin")
  end
end

platforms = Syro.new do
  @id = inbox.fetch(:id)

  get do
    res.write "GET /platforms/#{@id}"
  end
end

comments = Syro.new do
  get do
    res.write sprintf("GET %s/%s/comments",
      inbox[:path],
      inbox[:post_id])
  end
end

app = Syro.new do
  get do
    res.write "GET /"
  end

  post do
    on req.POST["user"] != nil do
      res.write "POST / (user)"
    end

    on true do
      res.write "POST / (none)"
    end
  end

  on "foo" do
    on "bar" do
      on "baz" do
        res.write("error")
      end

      get do
        res.write("GET /foo/bar")
      end

      put do
        res.write("PUT /foo/bar")
      end

      head do
        res.write("HEAD /foo/bar")
      end

      post do
        res.write("POST /foo/bar")
      end

      patch do
        res.write("PATCH /foo/bar")
      end

      delete do
        res.write("DELETE /foo/bar")
      end

      options do
        res.write("OPTIONS /foo/bar")
      end
    end
  end

  on "bar/baz" do
    get do
      res.write("GET /bar/baz")
    end
  end

  on "admin" do
    run(admin)
  end

  on "platforms" do
    run(platforms, id: 42)
  end

  on "rack" do
    run(RackApp.new)
  end

  on "users" do
    on :id do
      res.write(sprintf("GET /users/%s", inbox[:id]))
    end
  end

  on "posts" do
    @path = path.prev

    on :post_id do
      on "comments" do
        run(comments, inbox.merge(path: @path))
      end
    end
  end

  on "one" do
    @one = "1"

    get do
      res.write(@one)
    end
  end

  on "two" do
    get do
      res.write(@one)
    end

    post do
      res.redirect("/one")
    end
  end

  on "markdown" do
    run(markdown)
  end

  on "headers" do
    run(default_headers)
  end

  on "custom" do
    run(json)
  end

  on "private" do
    res.status = 401
    res.write("Unauthorized")
  end

  on "write" do
    res.write "nil!"
  end

  on "text" do
    res.text "plain!"
  end

  on "html" do
    res.html "html!"
  end

  on "json" do
    res.json "json!"
  end
end

setup do
  Driver.new(app)
end

test "path + verb" do |f|
  f.get("/foo/bar")
  assert_equal 200, f.last_response.status
  assert_equal "GET /foo/bar", f.last_response.body

  f.get("/bar/baz")
  assert_equal 404, f.last_response.status
  assert_equal "", f.last_response.body

  f.put("/foo/bar")
  assert_equal 200, f.last_response.status
  assert_equal "PUT /foo/bar", f.last_response.body

  f.head("/foo/bar")
  assert_equal 200, f.last_response.status
  assert_equal "HEAD /foo/bar", f.last_response.body

  f.post("/foo/bar")
  assert_equal 200, f.last_response.status
  assert_equal "POST /foo/bar", f.last_response.body

  f.patch("/foo/bar")
  assert_equal 200, f.last_response.status
  assert_equal "PATCH /foo/bar", f.last_response.body

  f.delete("/foo/bar")
  assert_equal 200, f.last_response.status
  assert_equal "DELETE /foo/bar", f.last_response.body

  f.options("/foo/bar")
  assert_equal 200, f.last_response.status
  assert_equal "OPTIONS /foo/bar", f.last_response.body
end

test "verbs match only on root" do |f|
  f.get("/bar/baz/foo")
  assert_equal "", f.last_response.body
  assert_equal 404, f.last_response.status
end

test "mounted app" do |f|
  f.get("/admin")
  assert_equal "GET /admin", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "mounted app + inbox" do |f|
  f.get("/platforms")
  assert_equal "GET /platforms/42", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "run rack app" do |f|
  f.get("/rack")
  assert_equal "GET /rack", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "root" do |f|
  f.get("/")
  assert_equal "GET /", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "captures" do |f|
  f.get("/users/42")
  assert_equal "GET /users/42", f.last_response.body

  # As the verb was not mached, the status is 404.
  assert_equal 404, f.last_response.status
end

test "post values" do |f|
  f.post("/", "user" => { "username" => "foo" })
  assert_equal "POST / (user)", f.last_response.body
  assert_equal 200, f.last_response.status

  f.post("/")
  assert_equal "POST / (none)", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "inherited inbox" do |f|
  f.get("/posts/42/comments")
  assert_equal "GET /posts/42/comments", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "leaks" do |f|
  f.get("/one")
  assert_equal "1", f.last_response.body
  assert_equal 200, f.last_response.status

  f.get("/two")
  assert_equal "", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "redirect" do |f|
  f.post("/two")
  assert_equal 302, f.last_response.status

  f.follow_redirect!
  assert_equal "1", f.last_response.body
  assert_equal 200, f.last_response.status
end

test "custom deck" do |f|
  f.get("/markdown")
  assert_equal "GET /markdown", f.last_response.body
  assert_equal "text/markdown", f.last_response.headers["Content-Type"]
  assert_equal 200, f.last_response.status
end

test "default headers" do |f|
  f.get("/headers")

  assert_equal "text/html", f.last_response.headers["Content-Type"]
end

test "custom request and response class" do |f|
  params = JSON.generate(foo: "foo")

  f.post("/custom", params)

  assert_equal params, f.last_response.body
end

test "don't set content type by default" do |f|
  f.get("/private")

  assert_equal 401, f.last_response.status
  assert_equal "Unauthorized", f.last_response.body
  assert_equal nil, f.last_response.headers["Content-Type"]
end

test "content type" do |f|
  f.get("/write")
  assert_equal nil, f.last_response.headers["Content-Type"]

  f.get("/text")
  assert_equal "text/plain", f.last_response.headers["Content-Type"]

  f.get("/html")
  assert_equal "text/html", f.last_response.headers["Content-Type"]

  f.get("/json")
  assert_equal "application/json", f.last_response.headers["Content-Type"]
end


test/helper.rb
require_relative "../lib/syro"
require "rack/test"

class Driver
  include Rack::Test::Methods

  def initialize(app)
    @app = app
  end

  def app
    @app
  end
end



lib/syro.rb
# encoding: UTF-8
#
# Copyright (c) 2015 Michel Martens
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

require "rack"
require "seg"

class Syro
  INBOX = "syro.inbox".freeze # :nodoc:

  class Response
    LOCATION = "Location".freeze # :nodoc:

    module ContentType
      HTML = "text/html".freeze        # :nodoc:
      TEXT = "text/plain".freeze       # :nodoc:
      JSON = "application/json".freeze # :nodoc:
    end

    # The status of the response.
    #
    #     res.status = 200
    #     res.status # => 200
    #
    attr_accessor :status

    # Returns the body of the response.
    #
    #     res.body
    #     # => []
    #
    #     res.write("there is")
    #     res.write("no try")
    #
    #     res.body
    #     # => ["there is", "no try"]
    #
    attr_reader :body

    # Returns a hash with the response headers.
    #
    #     res.headers
    #     # => { "Content-Type" => "text/html", "Content-Length" => "42" }
    #
    attr_reader :headers

    def initialize(headers = {})
      @status  = 404
      @headers = headers
      @body    = []
      @length  = 0
    end

    # Returns the response header corresponding to `key`.
    #
    #     res["Content-Type"]   # => "text/html"
    #     res["Content-Length"] # => "42"
    #
    def [](key)
      @headers[key]
    end

    # Sets the given `value` with the header corresponding to `key`.
    #
    #     res["Content-Type"] = "application/json"
    #     res["Content-Type"] # => "application/json"
    #
    def []=(key, value)
      @headers[key] = value
    end

    # Appends `str` to `body` and updates the `Content-Length` header.
    #
    #     res.body # => []
    #
    #     res.write("foo")
    #     res.write("bar")
    #
    #     res.body
    #     # => ["foo", "bar"]
    #
    #     res["Content-Length"]
    #     # => 6
    #
    def write(str)
      s = str.to_s

      @length += s.bytesize
      @headers[Rack::CONTENT_LENGTH] = @length.to_s
      @body << s
    end

    # Write response body as text/plain
    def text(str)
      @headers[Rack::CONTENT_TYPE] = ContentType::TEXT
      write(str)
    end

    # Write response body as text/html
    def html(str)
      @headers[Rack::CONTENT_TYPE] = ContentType::HTML
      write(str)
    end

    # Write response body as application/json
    def json(str)
      @headers[Rack::CONTENT_TYPE] = ContentType::JSON
      write(str)
    end

    # Sets the `Location` header to `path` and updates the status to
    # `status`. By default, `status` is `302`.
    #
    #     res.redirect("/path")
    #
    #     res["Location"] # => "/path"
    #     res.status      # => 302
    #
    #     res.redirect("http://syro.ru", 303)
    #
    #     res["Location"] # => "http://syro.ru"
    #     res.status      # => 303
    #
    def redirect(path, status = 302)
      @headers[LOCATION] = path
      @status = status
    end

    # Returns an array with three elements: the status, headers and
    # body. If the status is not set, the status is set to 404. If
    # a match is found for both path and request method, the status
    # is changed to 200.
    #
    #     res.status = 200
    #     res.finish
    #     # => [200, {}, []]
    #
    #     res.status = nil
    #     res.finish
    #     # => [404, {}, []]
    #
    #     res.status = nil
    #     res.write("syro")
    #     res.finish
    #     # => [200, { "Content-Type" => "text/html" }, ["syro"]]
    #
    def finish
      [@status, @headers, @body]
    end

    # Sets a cookie into the response.
    #
    #     res.set_cookie("foo", "bar")
    #     res["Set-Cookie"] # => "foo=bar"
    #
    #     res.set_cookie("foo2", "bar2")
    #     res["Set-Cookie"] # => "foo=bar\nfoo2=bar2"
    #
    #     res.set_cookie("bar", {
    #       domain: ".example.com",
    #       path: "/",
    #       # max_age: 0,
    #       # expires: Time.now + 10_000,
    #       secure: true,
    #       httponly: true,
    #       value: "bar"
    #     })
    #
    #     res["Set-Cookie"].split("\n").last
    #     # => "bar=bar; domain=.example.com; path=/; secure; HttpOnly
    #
    # **NOTE:** This method doesn't sign and/or encrypt the value of the cookie.
    #
    def set_cookie(key, value)
      Rack::Utils.set_cookie_header!(@headers, key, value)
    end

    # Deletes given cookie.
    #
    #     res.set_cookie("foo", "bar")
    #     res["Set-Cookie"]
    #     # => "foo=bar"
    #
    #     res.delete_cookie("foo")
    #     res["Set-Cookie"]
    #     # => "foo=; max-age=0; expires=Thu, 01 Jan 1970 00:00:00 -0000"
    #
    def delete_cookie(key, value = {})
      Rack::Utils.delete_cookie_header!(@headers, key, value)
    end
  end

  class Deck
    module API
      def initialize(code)
        @syro_code = code
      end

      def env
        @syro_env
      end

      # Returns the incoming request object. This object is an
      # instance of Rack::Request.
      #
      #     req.post?      # => true
      #     req.params     # => { "username" => "bob", "password" => "secret" }
      #     req[:username] # => "bob"
      #
      def req
        @syro_req
      end

      # Returns the current response object. This object is an
      # instance of Syro::Response.
      #
      #     res.status = 200
      #     res["Content-Type"] = "text/html"
      #     res.write("<h1>Welcome back!</h1>")
      #
      def res
        @syro_res
      end

      def path
        @syro_path
      end

      def inbox
        @syro_inbox
      end

      def default_headers
        {}
      end

      def request_class
        Rack::Request
      end

      def response_class
        Syro::Response
      end

      def call(env, inbox)
        @syro_env = env
        @syro_req = request_class.new(env)
        @syro_res = response_class.new(default_headers)
        @syro_path = Seg.new(env.fetch(Rack::PATH_INFO))
        @syro_inbox = inbox

        catch(:halt) do
          instance_eval(&@syro_code)

          @syro_res.finish
        end
      end

      def run(app, inbox = {})
        path, script = env[Rack::PATH_INFO], env[Rack::SCRIPT_NAME]

        env[Rack::PATH_INFO] = @syro_path.curr
        env[Rack::SCRIPT_NAME] = @syro_path.prev
        env[Syro::INBOX] = inbox

        halt(app.call(env))
      ensure
        env[Rack::PATH_INFO], env[Rack::SCRIPT_NAME] = path, script
      end

      # Immediately stops the request and returns `response`
      # as per Rack's specification.
      #
      #     halt([200, { "Content-Type" => "text/html" }, ["hello"]])
      #     halt([res.status, res.headers, res.body])
      #     halt(res.finish)
      #
      def halt(response)
        throw(:halt, response)
      end

      def consume(arg)
        @syro_path.consume(arg)
      end

      def capture(arg)
        @syro_path.capture(arg, inbox)
      end

      def root?
        @syro_path.root?
      end

      def match(arg)
        case arg
        when String then consume(arg)
        when Symbol then capture(arg)
        when true   then true
        else false
        end
      end

      def default
        yield; halt(res.finish)
      end

      def on(arg)
        default { yield } if match(arg)
      end

      def root
        default { yield } if root?
      end

      def get
        root { res.status = 200; yield } if req.get?
      end

      def put
        root { res.status = 200; yield } if req.put?
      end

      def head
        root { res.status = 200; yield } if req.head?
      end

      def post
        root { res.status = 200; yield } if req.post?
      end

      def patch
        root { res.status = 200; yield } if req.patch?
      end

      def delete
        root { res.status = 200; yield } if req.delete?
      end

      def options
        root { res.status = 200; yield } if req.options?
      end
    end

    include API
  end

  def initialize(deck = Deck, &code)
    @deck = deck
    @code = code
  end

  def call(env, inbox = env.fetch(Syro::INBOX, {}))
    @deck.new(@code).call(env, inbox)
  end
end
